File: fronted (deprecated)/src/main/java/com/sketchvision/fronted/algo/Matrix.java
package com.sketchvision.fronted.algo;

import java.util.Arrays;

public class Matrix {
    private final int rows;
    private final int cols;
    private final double[][] data;

    public Matrix(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.data = new double[rows][cols];
    }

    public static Matrix identity(int n) {
        Matrix m = new Matrix(n, n);
        for (int i = 0; i < n; i++) m.data[i][i] = 1.0;
        return m;
    }

    public static Matrix from(double[][] d) {
        int r = d.length;
        int c = r == 0 ? 0 : d[0].length;
        Matrix m = new Matrix(r, c);
        for (int i = 0; i < r; i++) m.data[i] = Arrays.copyOf(d[i], c);
        return m;
    }

    public Matrix multiply(Matrix other) {
        if (this.cols != other.rows) throw new IllegalArgumentException("shape mismatch");
        Matrix out = new Matrix(this.rows, other.cols);
        for (int i = 0; i < rows; i++) {
            for (int k = 0; k < cols; k++) {
                double a = this.data[i][k];
                for (int j = 0; j < other.cols; j++) {
                    out.data[i][j] += a * other.data[k][j];
                }
            }
        }
        return out;
    }

    @Override public String toString() { return Arrays.deepToString(data); }
}

File: frontend/internal/algo/matrix.go
package algo

import (
	"fmt"
	"strings"
)

type Matrix struct {
	Rows int
	Cols int
	Data [][]float64
}

func New(rows, cols int) Matrix {
	m := Matrix{Rows: rows, Cols: cols, Data: make([][]float64, rows)}
	for i := 0; i < rows; i++ {
		m.Data[i] = make([]float64, cols)
	}
	return m
}

func Identity(n int) Matrix {
	m := New(n, n)
	for i := 0; i < n; i++ { m.Data[i][i] = 1 }
	return m
}

func From(d [][]float64) Matrix {
	r := len(d)
	c := 0
	if r > 0 { c = len(d[0]) }
	m := New(r, c)
	for i := 0; i < r; i++ {
		copy(m.Data[i], d[i])
	}
	return m
}

func (m Matrix) Multiply(o Matrix) Matrix {
	if m.Cols != o.Rows {
		panic("shape mismatch")
	}
	out := New(m.Rows, o.Cols)
	for i := 0; i < m.Rows; i++ {
		for k := 0; k < m.Cols; k++ {
			a := m.Data[i][k]
			for j := 0; j < o.Cols; j++ {
				out.Data[i][j] += a * o.Data[k][j]
			}
		}
	}
	return out
}

func (m Matrix) String() string {
	rows := make([]string, m.Rows)
	for i := 0; i < m.Rows; i++ {
		rows[i] = fmt.Sprint(m.Data[i])
	}
	return "[" + strings.Join(rows, ", ") + "]"
}
